HistoryQueryManager/src/main/java/eu/neclab/ngsildbroker/historyquerymanager/service/HistoryQueryService.java

```diff
@@ -356,54 +373,106 @@ private Map<RemoteHost, Set<String>> getRemoteHostsForRetrieve(String tenant, St
+						filteredTypes.add(type);
+					}
 				}
 			}
+			else{
+				filteredTypes=types;
+			}
+
+                queryMap.put("id",String.join(",",filteredIds));
+				queryMap.put("type",String.join(",",filteredTypes));
+
+            if (regEntry.eProp() != null || regEntry.eRel() != null) {
+                Set<String> uniqueAttrs = new HashSet<>( Arrays.asList(queryMap.getOrDefault("attrs","").split(",")));
+                uniqueAttrs.add(regEntry.eProp());
+                uniqueAttrs.add(regEntry.eRel());
+				uniqueAttrs.remove("");
+				uniqueAttrs.remove(null);
+                queryMap.put("attrs",String.join(",",uniqueAttrs));
+            }
+			result.put(remoteHost,queryMapToStr(queryMap));
 		}
-//		if (attrsQuery != null) {
-//			for (Entry<RemoteHost, Set<String>> entry : result.entrySet()) {
-//				Set<String> regAttrs = entry.getValue();
-//				if (regAttrs.isEmpty()) {
-//					regAttrs.addAll(attrsQuery.getAttrs());
-//				} else {
-//					regAttrs.retainAll(attrsQuery.getAttrs());
-//				}
-//			}
-//		}
 		return result;
 	}
+	private static Map<String, String> queryStrToMap(String queryString) {
+		Map<String, String> paramMap = new HashMap<>();
+
+		String[] paramPairs = queryString.split("&");
+		for (String paramPair : paramPairs) {
+			String[] keyValue = paramPair.split("=");
+			if (keyValue.length == 2) {
+				String key = keyValue[0];
+				String value = keyValue[1];
+				paramMap.put(key, value);
+			}
+		}
+
+		return paramMap;
+	}
+	public static String queryMapToStr(Map<String, String> paramMap) {
+		StringBuilder queryStringBuilder = new StringBuilder();
+		List<String> keyList = paramMap.keySet().stream().toList();
+        for (String key : keyList) {
+            if (paramMap.get(key).isEmpty()) {
+                paramMap.remove(key);
+            }
+        }
+
+		for (Map.Entry<String, String> entry : paramMap.entrySet()) {
+			if (!queryStringBuilder.isEmpty()) {
+				queryStringBuilder.append("&");
+			}
+			try {
+				String key = URLEncoder.encode(entry.getKey(), StandardCharsets.UTF_8);
+				String value = URLEncoder.encode(entry.getValue(), StandardCharsets.UTF_8);
+				queryStringBuilder.append(key).append("=").append(value);
+			} catch (Exception ignored) {}
+		}
+
+		return queryStringBuilder.toString();
+	}
 
 	public Uni<Void> handleRegistryChange(BaseRequest req) {
 		tenant2CId2RegEntries.remove(req.getTenant(), req.getId());
 		if (req.getRequestType() != AppConstants.DELETE_REQUEST) {
-			for (RegistrationEntry regEntry : RegistrationEntry.fromRegPayload(req.getPayload())) {
+			List<RegistrationEntry> lsReg = RegistrationEntry.fromRegPayload(req.getPayload());
+			int count=0;
+			for (RegistrationEntry regEntry : lsReg) {
 				if (regEntry.retrieveTemporal() || regEntry.queryTemporal()) {
-					tenant2CId2RegEntries.put(req.getTenant(), req.getId(), regEntry);
+					tenant2CId2RegEntries.put(req.getTenant(), req.getId()+count++, regEntry);
```

Comment:
please change this. this is supposed to overwrite the entries on changes. i know it was not properly working to begin with like this but please have a look at the entity service for instance and do it like this thanks